src/librustdoc/visit_ast.rs

```diff
@@ -79,50 +79,25 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
-        // Attach the crate's exported macros to the top-level module.
-        // In the case of macros 2.0 (`pub macro`), and for built-in `derive`s or attributes as
-        // well (_e.g._, `Copy`), these are wrongly bundled in there too, so we need to fix that by
-        // moving them back to their correct locations.
-        'exported_macros: for def in krate.exported_macros() {
-            // The `def` of a macro in `exported_macros` should correspond to either:
-            //  - a `#[macro_export] macro_rules!` macro,
-            //  - a built-in `derive` (or attribute) macro such as the ones in `::core`,
-            //  - a `pub macro`.
-            // Only the last two need to be fixed, thus:
-            if def.ast.macro_rules {
-                top_level_module.macros.push((def, None));
-                continue 'exported_macros;
-            }
-            let tcx = self.cx.tcx;
-            // Note: this is not the same as `.parent_module()`. Indeed, the latter looks
-            // for the closest module _ancestor_, which is not necessarily a direct parent
-            // (since a direct parent isn't necessarily a module, c.f. #77828).
-            let macro_parent_def_id = {
-                use rustc_middle::ty::DefIdTree;
-                tcx.parent(def.def_id.to_def_id()).unwrap()
-            };
-            let macro_parent_path = tcx.def_path(macro_parent_def_id);
-            // HACK: rustdoc has no way to lookup `doctree::Module`s by their HirId. Instead,
-            // lookup the module by its name, by looking at each path segment one at a time.
-            let mut cur_mod = &mut top_level_module;
-            for path_segment in macro_parent_path.data {
-                // Path segments may refer to a module (in which case they belong to the type
-                // namespace), which is _necessary_ for the macro to be accessible outside it
-                // (no "associated macros" as of yet). Else we bail with an outer `continue`.
-                let path_segment_ty_ns = match path_segment.data {
-                    rustc_hir::definitions::DefPathData::TypeNs(symbol) => symbol,
-                    _ => continue 'exported_macros,
-                };
-                // Descend into the child module that matches this path segment (if any).
-                match cur_mod.mods.iter_mut().find(|child| child.name == path_segment_ty_ns) {
-                    Some(child_mod) => cur_mod = &mut *child_mod,
-                    None => continue 'exported_macros,
+
+        // `#[macro_export] macro_rules!` items are reexported at the top level of the
+        // crate, regardless of where they're defined. We want to document the
+        // top level rexport of the macro, not its original definition, since
+        // the rexport defines the path that a user will actually see. Accordingly,
+        // we add the rexport as an item here, and then skip over the original
+        // definition in `visit_item()` below.
```

Comment:
So, if non-exported `macro_rules` are documented at their in-source definition point when `--document-private-items` is enabled, then why exported `macro_rules` are not? :)

It would be pretty reasonable/consistent to document exported macros twice in the doc-private mode - once as an item and once as the root "reexport" (especially if that would make rustdoc logic simpler).

Reply:
Documenting them twice feels weird to me, especially given that the only path they can be addressed at is the root path. Would it be worth asking the rustdoc team?

Reply:
@petrochenkov I don't understand your suggestion. Why would rustdoc document macros at the definition if they're not accessible using that path? That seems actively misleading.

Reply:
>given that the only path they can be addressed at is the root path

Non-exported macros cannot be addressed by any path, but they are still documented when `--document-private-items` is enabled.
This comment is certainly not a blocking one though.

Reply:
> Non-exported macros cannot be addressed by any path, but they are still documented when --document-private-items is enabled.

Can't they be addressed by that path by the module that defined them?

Reply:
No, `macro_rules` macros are not "modularized" and cannot be addressed by a module-relative paths.
Moreover, there can be multiple `macro_rules` with the same name in the same module.

Reply:
> Moreover, there can be multiple macro_rules with the same name in the same module.

Oh fun ... Rustdoc almost certainly can't handle that. @inquisitivecrystal can you add a test for that case? I expect all the macro.name.html files will be overwritten by each other.